fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2954,6 +2956,121 @@ public void validateMetaData() {
+                    return true;
+                })).thenApply(ignored -> nextContinuation.get());
+    }
+
+    private void repairRecordKeyIfNecessary(@Nonnull FDBRecordContext context, @Nonnull Subspace recordSubspace, @Nonnull KeyValue keyValue) {
+        final RecordMetaData metaData = metaDataProvider.getRecordMetaData();
+        final Tuple recordKey = recordSubspace.unpack(keyValue.getKey());
+
+        // Ignore version key
+        if (metaData.isStoreRecordVersions() && isMaybeVersion(recordKey)) {
+            return;
+        }
+
+        final Message record = serializer.deserialize(metaData, recordKey, keyValue.getValue(), getTimer());
+        final RecordType recordType = metaData.getRecordTypeForDescriptor(record.getDescriptorForType());
+
+        final KeyExpression primaryKeyExpression = recordType.getPrimaryKey();
+        if (!primaryKeyExpression.isFixedColumnSize()) {
+            throw new RecordCoreException("Cannot repair records with variable length primary keys")
+                    .addLogInfo(
+                            subspaceProvider.logKey(), subspaceProvider.toString(context),
+                            LogMessageKeys.RECORD_TYPE, recordType.getName(),
+                            LogMessageKeys.KEY_EXPRESSION, primaryKeyExpression);
+        }
+
+        if (recordKey.size() == primaryKeyExpression.getColumnSize()) {
+            context.increment(FDBStoreTimer.Counts.REPAIR_RECORD_KEY);
+
+            final Tuple newPrimaryKey = recordKey.add(SplitHelper.UNSPLIT_RECORD);
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug(KeyValueLogMessage.of("Repairing primary key",
+                        LogMessageKeys.RECORD_TYPE, recordType.getName(),
+                        subspaceProvider.logKey(), subspaceProvider.toString(context),
+                        "orig_primary_key", recordKey,
+                        "new_primary_key", newPrimaryKey));
+            }
+
+            final Transaction tr = context.ensureActive();
+            tr.clear(keyValue.getKey());
+            tr.set(recordSubspace.pack(newPrimaryKey), keyValue.getValue());
+        } else if (recordKey.size() == primaryKeyExpression.getColumnSize() + 1) {
+            Object suffix = recordKey.get(recordKey.size() - 1);
+            if (!(suffix instanceof Long) || !(((Long) suffix) == SplitHelper.UNSPLIT_RECORD)) {
+                context.increment(FDBStoreTimer.Counts.INVALID_SPLIT_SUFFIX);
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(KeyValueLogMessage.of("Invalid split suffix",
+                            subspaceProvider.logKey(), subspaceProvider.toString(context),
+                            LogMessageKeys.RECORD_TYPE, recordType.getName(),
+                            LogMessageKeys.PRIMARY_KEY, recordKey));
+                }
+            }
+        } else if (recordKey.size() > primaryKeyExpression.getColumnSize() + 1) {
```

Comment:
Why not just `else`? What is the remaining legit case?

Reply:
Good point. Fixed.